README
Speak is in intermediate state of switching to 0sugar, "master" branch code
should work as-is but requires toolkit dependency, grab it from "fully-bundled"
branch.


NEWS
59

* Fix saving photo (Martin Abente Lahaye),
* New translations (Chris Leonard et al),
* Refresh POT file (James Cameron),

58

* Replace screenshots and update metadata (Ashnidh),
* Port to Telepathy via D-Bus (James Cameron),
* Port to Python 3 (Ibiam Chihurumnaya),
* Update AIML and recompile Brain (Ibiam Chihurumnaya),
* Fix missing sunglasses (James Cameron),
* Port to sugar3.speech (Ibiam Chihurumnaya),
* Fix missing mouth on welcome speech (James Cameron),
* Fix support for non-english languages (Martin Abente Lahaye),

57

* Port from GConf to Gio.Settings (James Cameron),
* New translations (Chris Leonard et al),
* Port from GObject to GLib (James Cameron),
* Fix occasional ValueError (James Cameron),
* Fix occasional ZeroDivisionError (James Cameron),
* Port to TelepathyGLib (James Cameron),

56

* Fix fail to start when used with espeak-ng 1.49.2 (James Cameron),
* New translations (Chris Leonard et al),

55

* Fix fail to start when used with espeak-ng 1.49.1 (James Cameron),
* New translations (Chris Leonard et al),
* Update POT file (James Cameron),

54

* Remove curse filter; ineffective, untranslated and is offensive
  source code (James Cameron),
* New translations (Chris Leonard et al),
* Update POT file (James Cameron),

53

* New translations
* Fix repository in activity.info
* Port to GTK 3 and GStreamer 1.0
* Make mouth animation follow sound more closely
* Enable URL clicking in chat view
* Fix accelerometer fun (v48)
* Fix swap of waveform and frequency mouth buttons

52

* New translations
* Added repository to activity.info

51

* Compatibity change for Debian (Gonzalo Odiard)

50

* New icon for photoface

49

* Added talking photoface (samdroid with an assist from ignacio)

48

* Added idle-time chatter
* Added some accelerometer fun
* Added more eye styles
* Using XO color for background
* Redesigned chat look and feel
* Using palettes instead of radio buttons
* Using palettes instead of combo boxes
* Removed deprecated code

41

* Fix issue of not packaging .mo translation files
* Update translations

40

* Add DescriptionItem to the toolbar.

39

* Update translations

38

* Update translations

37

* Fixed combobox for voices, reduced to 28 chars max (Ariel Calzada)
* Update translations

36

* Fix pronunciation of unicode strings; keep history in unicode (Gonzalo Odiard)
* Update translations

35

* Update translations

34

*Fix i18n warning (Aleksey Lim)

33

*New translations

32

* Set-english-rp-as-default-for-en_AU (Sebastian Silva)
* avoiding gettext warning
* New translations

31

* Remove KeepButton due to deprecation (Gonzalo Odiard)
* Update i18n

30

* minor typo fix

29

* Fixing typo in LANG name

28

*Adding friendlynames for english_rp and english_wmids, english-us accents.
*various pep8 and pylint fixes.
*updated translations

26

*Enabling english_rp and english_wmids accents

25

*Adding AUTHORS file
*pep8 fixes
*Being more polite
*Adding new answers
*New pootle templates

24

*Fixing SL#2703

23

* Installing as speak.activity as default not as hablarconsara 

22

* Fixing  #2630 b.sl.o and d.l.o #10666

21

* Speak's robot is missing a some nouns in os439dg on XO-1.5 #2566
* Switch face voice to known language for robot on fallback
* More AIML tweaks

20

* Translate only first part of voice names (ie language name)
* Use short voice names while choosing Robot tab language
* Update es translation

19

* talk to computer crashes #2533
* Use the same default rate in espeak command mode like in gst mode #2182
* Update ta translation

18

* Remove content from AIML sets that is inappropriate to education

17

* Get rid of "smart" bundles(for ASLO).. and look ahead for smart packaging system
* Add copyright notes to python files

16

* Fix launching under rainbow

15

* Switch to restricted 0sugar mode until more complex solution is accessible,
  like Saccharin sugar distribution
* Traslation fixes

14

* Workaround for #1801
* Update vi translation

13

* Speak dies in robot mode #1779
* Update fr translation

12

* Add toolkit service to use new toolbars design in all supported sugars
* Add gst-plugins-espeak service to use gst in 0.82
* Speak offline mode warning
* Do not cache bot brains to save memory
* Load standard AIML set for restricted systems
* Implant xeyes

11

* Pronounce gettext strings in native language
* Load brain after choosing voice from combobox in "Robot" tab
* Add ru.po

10

* Merge HabarConSara branch
* Add "Robot" tab to use HabarConSara functionality
* #743: FIx crash for as sintaxis (dirakx)

7

* #7852: Fix crash on launch by using numpy (morgs)


espeak.py
# Copyright (C) 2009, Aleksey Lim
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

import numpy

from gi.repository import Gst
from gi.repository import GLib
from gi.repository import GObject

import logging
logger = logging.getLogger('speak')

PITCH_MIN = 0
PITCH_MAX = 200
RATE_MIN = 0
RATE_MAX = 200


class BaseAudioGrab(GObject.GObject):
    __gsignals__ = {
        'peak': (GObject.SIGNAL_RUN_FIRST, None, [GObject.TYPE_PYOBJECT]),
        'wave': (GObject.SIGNAL_RUN_FIRST, None, [GObject.TYPE_PYOBJECT]),
        'idle': (GObject.SIGNAL_RUN_FIRST, None, []),
    }

    def __init__(self):
        GObject.GObject.__init__(self)
        self.pipeline = None

        self._cb = {}
        for cb in ['peak', 'wave', 'idle']:
            self._cb[cb] = None

    def disconnect_all(self):
        for cb in ['peak', 'wave', 'idle']:
            hid = self._cb[cb]
            if hid is not None:
                self.disconnect(hid)
                self._cb[cb] = None

    def connect_peak(self, cb):
        self._cb['peak'] = self.connect('peak', cb)

    def connect_wave(self, cb):
        self._cb['wave'] = self.connect('wave', cb)

    def connect_idle(self, cb):
        self._cb['idle'] = self.connect('idle', cb)

    def restart_sound_device(self):
        self.pipeline.set_state(Gst.State.NULL)
        self.pipeline.set_state(Gst.State.PLAYING)

    def stop_sound_device(self):
        if self.pipeline is None:
            return

        self.pipeline.set_state(Gst.State.NULL)
        self.emit("idle")

        self.pipeline = None

    def make_pipeline(self):
        if self.pipeline is not None:
            self.stop_sound_device()
            del self.pipeline

        # build a pipeline that makes speech
        # and sends it to both the audio output
        # and a fake one that we use to draw from
        cmd = 'espeak name=espeak' \
            ' ! capsfilter name=caps' \
            ' ! tee name=me' \
            ' me.! queue ! autoaudiosink name=ears' \
            ' me.! queue ! fakesink name=sink'
        self.pipeline = Gst.parse_launch(cmd)

        # force a sample bit width to match our numpy code below
        caps = self.pipeline.get_by_name('caps')
        want = 'audio/x-raw,channels=(int)1,depth=(int)16'
        caps.set_property('caps', Gst.caps_from_string(want))

        # grab reference to the output element for scheduling mouth moves
        ears = self.pipeline.get_by_name('ears')

        def handoff(element, data, pad):
            size = data.get_size()
            if size == 0 or data.duration == 0:
                return True  # common

            npc = 50000000  # nanoseconds per chunk
            bpc = size * npc // data.duration  # bytes per chunk
            bpc = bpc // 2 * 2  # force alignment for int16

            a = []
            p = []
            w = []

            here = 0  # offset in bytes
            when = data.pts
            last = data.pts + data.duration
            while True:
                wave = numpy.fromstring(data.extract_dup(here, bpc), 'int16')
                peak = numpy.core.max(wave)

                a.append(wave)
                p.append(peak)
                w.append(when)

                here += bpc
                when += npc
                if when < last:
                    continue
                break

            def poke(pts):
                success, position = ears.query_position(Gst.Format.TIME)
                if not success:
                    return False

                if len(w) == 0:
                    return False

                if position < w[0]:
                    return True

                self.emit("wave", a[0])
                self.emit("peak", p[0])
                del a[0]
                del w[0]
                del p[0]

                if len(w) > 0:
                    return True

                return False

            GLib.timeout_add(25, poke, data.pts)

            return True

        sink = self.pipeline.get_by_name('sink')
        sink.props.signal_handoffs = True
        sink.connect('handoff', handoff)

        def gst_message_cb(bus, message):
            self._was_message = True

            if message.type == Gst.MessageType.WARNING:
                def check_after_warnings():
                    if not self._was_message:
                        self.stop_sound_device()
                    return True

                logger.debug(message.type)
                self._was_message = False
                GLib.timeout_add(500, check_after_warnings)

            elif message.type in (Gst.MessageType.EOS, Gst.MessageType.ERROR):
                logger.debug(message.type)
                self.stop_sound_device()
            return True

        self._was_message = False
        bus = self.pipeline.get_bus()
        bus.add_signal_watch()
        bus.connect('message', gst_message_cb)


class AudioGrab(BaseAudioGrab):
    def speak(self, status, text):
        self.make_pipeline()
        src = self.pipeline.get_by_name('espeak')

        pitch = int(status.pitch) - 100
        rate = int(status.rate) - 100

        logger.debug('pitch=%d rate=%d voice=%s text=%s' % (pitch, rate,
                                                            status.voice.name,
                                                            text))

        src.props.pitch = pitch
        src.props.rate = rate
        src.props.voice = status.voice.name
        src.props.track = 1
        src.props.text = text

        self.restart_sound_device()


def voices():
    out = []

    for i in Gst.ElementFactory.make('espeak', 'espeak').props.voices:
        name, language, dialect = i
        out.append((language, name))

    return out
